### ISSUE TITLE
"Window Placement Centered" causes game window to appear offscreen on hi-DPI laptop

### ISSUE THREAD

**Issue by zejji** [godotengine/godot#13017]
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

**Operating system or device, Godot version, GPU Model and driver (if graphics related):**

OS: Windows 10
Device: Dell XPS 15 9560
GPU: Nvidia GTX 1050
GPU Driver: 388.00
Screen Resolution:  3840 x 2160 px
Godot version: 3.0 alpha 2 

**Issue description:**

- Actual behaviour:

If Window Placement - Centred is selected in Editor Settings, when running a project using the 'Play Project' button, the game window opens outside the visible screen area (to the bottom-right). I have to move the window back onscreen to see it.

(However, Window Placement - Top Left works fine.)

- Expected behaviour:

The game window should be centred if Window Placement - Centred is selected in Editor Settings.

**Steps to reproduce:**

Load any project which runs windowed and ensure that Window Placement - Centred is selected in Editor Settings. Press 'Play Project'. 

**Link to minimal example project:**

*Bugsquad edit:* Keywords for searching: Windows, window, placement, location, position, screen, scaling, incorrect, offscreen, bottom, right, corner, bottom-right, center, centered, 125%, 150%, 175%, 200%, expected, unexpected, resolution.

**Comment by Calinou** [comment]
Does this happen if you mark the editor executable as DPI-aware, by right-clicking its `.exe` file, going to **Properties → Compatibility** then overriding the scaling mode to **Application**?

**Comment by zejji** [comment]
@Calinou - many thanks for the quick response.

I can confirm that overriding the scaling mode to 'Application' does solve the issue, i.e. the window which appears on pressing 'Play Project' is now centred.

One minor side effect which I noticed is that, after switching to 'Application' scaling mode, the Project Manager window which loads on startup is now a bit too small by default (a screenshot is attached).

![screenshot](https://user-images.githubusercontent.com/3524191/32986259-568e0256-ccc6-11e7-8852-f9ab400275a0.jpg)

**Comment by Calinou** [comment]
> One minor side effect which I noticed is that, after switching to 'Application' scaling mode, the Project Manager window which loads on startup is now a bit too small by default (a screenshot is attached).

[I have already reported that](https://github.com/godotengine/godot/issues/8543), so I'm adding the "confirmed" label to that issue.

**Comment by willvincent** [comment]
This is not limited to windows. I see the same on my mid 2014 macbook pro with retina display
<img width="1920" alt="screen shot 2017-12-11 at 12 51 50 am" src="https://user-images.githubusercontent.com/689891/33819053-f1faf0ac-de0d-11e7-91ad-4892c78baf37.png">

<img width="1920" alt="screen shot 2017-12-11 at 12 55 53 am" src="https://user-images.githubusercontent.com/689891/33819072-17aa3e5c-de0e-11e7-9237-9d3d859d3422.png">

**Comment by richardeigenmann** [comment]
Motivated by the presentations at FOSDEM 2018 I downloaded Godot 3.0-stable on my Win 10 Lenovo Yoga 3 Pro hidpi laptop and started the getting started tutorial. It was most confusing to not find the play window despite repeated play button pushes. Only after some head-scratching and right-clicking on the taskbar icons did it become apparent that the window is somewhere far off to the bottom right of the desktop and needs to be dragged up with "Move" or "Maximise". This is not a good experience to the newbie trying to get started.

**Comment by Arkhano** [comment]
I confirm this issue. In a 1920x1080 screen I need to put custom position like (420, 130) to the run screen be "centered".

**Comment by jaran** [comment]
I have this issue on mac.  took me forever to finally see a small sliver of the window bottom right.

**Comment by willvincent** [comment]
I no longer have this issue on 3.0.2

**Comment by Nachasic** [comment]
OS X 10.13.4
MacBook Pro with Retina with external FullHD monitor (which Godot is running on), same issue

**Comment by cephalopodluke** [comment]
Same problem here. :/

**Comment by ffxsam** [comment]
Still having this problem in the latest version of Godot (3.0.2). macOS 10.13.4.

**Comment by KoBeWi** [comment]
I guess that would fix it: #23203

**Comment by ReneHabermann** [comment]
Can reproduce at 3440x1440 resolution. Apparently it's not related to the windows UI scaling, same offset for 100% scaling and 150% scaling.

**Comment by rolbr** [comment]
Same problem here. Setting `Allow HiDPI` in the project settings kinda fixes this issue and correctly centers the game window, but then it'll be super tiny and you need to also set `Test Width` and `Test Height` to something more manageable.

Leaving `Allow HiDPI` **disabled** but then set `Editor Settings > Run > Window Placement > Rect` to `Custom Position` and giving the rect a position of `248,100` also centers the game window correctly — but only for that very project resolution.

System:
  - Win 10 on a Dell XPS 9555
  - 3840×2160
  - window scaling 250% (default)

Project Settings: 
  - 1024 × 600
  - HiDPI disabled

Editor Settings:
  - `Window Placement` custom rect at 248, 100

Result:
| | margin left | game window width | margin right | total |
|---|---|---|---|---|
|Device PX| 640 | 2560 | 640 | 3840 |
|divided by 2.5 | 256 | **1024** | 256 | 1536 |

The first row are the pixel values measured from a screenshot when running the game (I only measured the X axis, but I'm pretty sure vertically it's the same)
The second row then is all the values from the top row divided by 2.5 as per the 250% window scaling.

Soooo... the measured window width of 2560 perfectly translates to the project setting of 1024, I guess the Editor Setting `Window Placement` somewhere mixes up the scaling between game pixels and device pixels.

**Comment by HubKing** [comment]
I was about to open a new issue for this, but since it already exists... I am experiencing the same problem. 

Godot Version: 3.2.1.stable.mono.official
OS: Windows
Monitor: 4K, 200DPI.

When the position is set to "Centred", it shows only the second quadrant of the game window at the bottom-right corner of the monitor.

**Comment by Calinou** [comment]
@HubKing Try enabling **Display > Window > Allow Hidpi** in the Project Settings.

**Comment by HubKing** [comment]
@Calinou That did make the game window appear at the centre. Thank you for the information. It may be helpful to show some kind of information message box like "Warning: high DPI option has not been enabled" (with "Do not show again" option, of course) when the user clicks the "Play" button for the first time and the system uses high DPI, so that the user would not think it as a bug of the editor.

**Comment by KoBeWi** [comment]
> so that the user would not think it as a bug of the editor.

This IS a bug of the editor though and an annoying one, especially when you run lots of different projects and you have to set this option for each of them.

**Comment by KoBeWi** [comment]
Reopening, because it wasn't fixed (the PR was for Mac OS, while this is also Windows issue, maybe that's why).
Tested in 8ccb1cec9

**Comment by eka** [comment]
> @HubKing Try enabling **Display > Window > Allow Hidpi** in the Project Settings. (For the record, this isn't done by default because your game must be configured to support [multiple resolutions](https://docs.godotengine.org/en/latest/tutorials/viewports/multiple_resolutions.html). Also, this may perform poorly on integrated graphics.)

This fixed my problem and the window appears centered

**Comment by SirBrott** [comment]
Im having the same problem
going to exe properties > compatibility > change high DPI settings and setting the High DPI scale override to application solved this problem, 
also setting hiDPI in the project solved this issue

OS: Windows 10
Device: PC
GPU: Nvidia GTX 970
GPU Driver: 452.06
Screen Resolution: 3840 x 2160 px
Godot version: 3.2.1 stable

**Comment by Calinou** [comment]
To fix this, we either need to:

- Find a way to set a window position that accounts for the OS' scaling in the Windows API.
- Add a method that returns the OS' scaling factor in Godot (not the DPI!). 1.0 would be 100%, 2.0 would be 200% and so on. We can then use this to position the window correctly.

**Comment by bruvzg** [comment]
> Add a method that returns the OS' scaling factor in Godot (not the DPI!). 1.0 would be 100%, 2.0 would be 200% and so on. We can then use this to position the window correctly.

`OS`/`DisplayServer`.`get_screen_scale` method exists, but it is only implemented on macOS.

On Windows Godot currently use old DPI model (`SHC_PROCESS_SYSTEM_DPI_AWARE`) which do not allow getting scale for specific monitor, and if it's changed to new model (`SHC_PROCESS_PER_MONITOR_DPI_AWARE`) we'll need to implement additional interface rescaling on screen change similar to one used on macOS or proper DPI independent UI for all platforms.

**Comment by Feniks-Gaming** [comment]
In 3.2.3 this is still an issue @ffxsam solution still works "Though if I go to Project Settings > Display > Window and enable "Allow Hidpi", this fixes it." but it's PITA to have to do it for each project.

**Comment by KoBeWi** [comment]
This is also an issue on master. The interesting part is that the window appears at correct position until it loads and then it moves to wrong position xd
![J3p4MHrk4l](https://user-images.githubusercontent.com/2223172/101375260-764b2300-38af-11eb-9f57-01807db623b9.gif)

**Comment by sarchar** [comment]
There's a couple issues here that I think are slightly separate.

(currently referring to Godot 3.3.stable.official)

1) The Project Manager on my 4K laptop display starts offscreen on the bottom-right but looks properly scaled.  It's just off screen. Changing compatibility mode to Application didn't seem to do anything for me.  Same effect.

2) The Editor itself also has this problem, but it's not immediately apparent because the Editor starts maximized; however, if you normalize the window it'll slide offscreen too.

3) Any game run from within the editor (with Allow HiDPI off) starts off screen.   Turning Allow HiDPI on and setting the window size to 1024x768 correctly centers the window, however all the UI elements are small and not ideal for high resolution monitors -- this is to be expected as the monitor is high DPI and the window is small.  So setting Test Width/Height to 2048x1536 just makes a bigger window and the UI elements' sizes stay the same.  Enabling Stretch=2d or 2d viewport with aspect ratio of keep makes the UI elements look normal, window still centered properly.  However, exporting the build and running it on a non-hiDPI monitor makes things look huge and starts off screen on a 1080p monitor!  A single build really should work for both monitors.

Note that for #1 and #2 all the UI elements do appear to be correctly scaled, but I'm not entirely sure what they look like on a 1080p monitor for comparison.

I don't know if the above provides new information or not:)

**Comment by Calinou** [comment]
> Fair, but what's the point of Test Width if it's being included in the exported build? The word Test sort of implies it's only for testing -- as if it won't be included in the export.

See https://github.com/godotengine/godot/pull/36390.

**Comment by Calinou** [comment]
> What do you think. Can we fix this soon? I could, but not experienced enough to this code. But can help with some ideas. Can we just use the same implementation as the "Allow HiDPI" for this launcher window?

**Allow Hidpi** is always enabled for the editor and project manager unless you pass the `--low-dpi` command line argument.

**Comment by twoco** [comment]
Ok thanks. `C:\...\Godot.exe resolution WxH` works in a Windows shortcut. Good to know. 👍 

> Allow Hidpi is always enabled for the editor and project manager unless you pass the --low-dpi command line argument.

It's only about the project manager who isn't working (position only, scale works). But `resolution WxH` works. In my case this is what I wanted. The project manager is not centered and the editor and game works like before. I just had to enable "Allow HiDPI" for my game. And for me I also set the Test Width / Height to have an internal resolution what modern engines has. ... Anyway, we could set this parameter by default. But what exactly does this mean (WxH)? Any docs?

**Comment by Calinou** [comment]
> It's only about the project manager who isn't working (position only, scale works).

This is because the code only sets the window size, not the position: 

https://github.com/godotengine/godot/blob/1ceba6e1b0c6925fd03e251f2197170ba5079ec6/editor/project_manager.cpp#L2415

I'll look into it but I can't promise I'll manage to fix it.

**Comment by twoco** [comment]
But does that mean there is code for the window position that won't execute when this error is caused? In other words: We can just remove the section of window position code to fix this issue? Or what do you think? Why is the window centered in this case? 

Not this line:

https://github.com/godotengine/godot/blob/1ceba6e1b0c6925fd03e251f2197170ba5079ec6/editor/project_manager.cpp#L2413-L2414

Because the widnow size is exactly the same as without the error / invalid argument. It's really only the position which is now fixed / centered. Or do you think that this is the affected line for this? But the window size is not different. 🤔

**Comment by Calinou** [comment]
> So just removing it will fix this? (Maybe not for all users?)

Removing that line will cause the window to be too small since it won't be resized to match the editor scale.

**Comment by cybereality** [comment]
So I spent the day trying to detect display scaling in Linux (Ubuntu Wayland and X11). It doesn't look like an easy thing. I was able to get some resolution values from the Xlib library, but they don't seem to make any sense. I tried everything I could think of, but it just wasn't working. I'll revisit this later because there has to be some way, maybe I'm missing something. I don't know.

**Comment by cybereality** [comment]
Actually, I was able to get it working with Xlib. At least on Ubuntu X11 GNOME. I can detect the proper screen scaling now. However, it doesn't seem to work under XWayland and I scoured the whole API. I also think that different desktop environments handle scaling differently, so I'm not sure it's portable to other distros. But it's a start and I will keep looking into it.

**Comment by sabbir360** [comment]
> Same problem here. Setting `Allow HiDPI` in the project settings kinda fixes this issue and correctly centers the game window, but then it'll be super tiny and you need to also set `Test Width` and `Test Height` to something more manageable.
> 
> Leaving `Allow HiDPI` **disabled** but then set `Editor Settings > Run > Window Placement > Rect` to `Custom Position` and giving the rect a position of `248,100` also centers the game window correctly — but only for that very project resolution.
> 
> System:
> 
> * Win 10 on a Dell XPS 9555
> * 3840×2160
> * window scaling 250% (default)
> 
> Project Settings:
> 
> * 1024 × 600
> * HiDPI disabled
> 
> Editor Settings:
> 
> * `Window Placement` custom rect at 248, 100
> 
> Result:
> 
> margin left	game window width	margin right	total
> Device PX	640	2560	640	3840
> divided by 2.5	256	**1024**	256	1536
> The first row are the pixel values measured from a screenshot when running the game (I only measured the X axis, but I'm pretty sure vertically it's the same) The second row then is all the values from the top row divided by 2.5 as per the 250% window scaling.
> 
> Soooo... the measured window width of 2560 perfectly translates to the project setting of 1024, I guess the Editor Setting `Window Placement` somewhere mixes up the scaling between game pixels and device pixels.

It's working for 4k display with scaling at 250%
